infrastructure/src/main/java/org/corfudb/infrastructure/orchestrator/actions/RestoreRedundancyMergeSegments.java

```diff
@@ -41,35 +176,80 @@ public String getName() {
         return "RestoreRedundancyAndMergeSegments";
     }
 
+    @FunctionalInterface
+    public interface LogUnitClientMap {
+        Map<String, LogUnitClient> generate(CorfuRuntime runtime);
+    }
+
     @Override
-    public void impl(@Nonnull CorfuRuntime runtime) throws Exception {
+    public void impl(@Nonnull CorfuRuntime runtime, @NonNull StreamLog streamLog) throws Exception {
 
         // Refresh layout.
         runtime.invalidateLayout();
         Layout layout = runtime.getLayoutView().getLayout();
 
-        // Each segment is compared with the first segment. The data is restored in any new LogUnit nodes and then
-        // merged to this segment. This is done for all the segments.
-        final int layoutSegmentToMergeTo = 0;
+        // Create a helper to perform the state calculations.
+        PrefixTrimRedundancyCalculator redundancyCalculator =
+                new PrefixTrimRedundancyCalculator(currentNode, runtime);
 
-        // Catchup all servers across all segments.
-        while (layout.getSegments().size() > 1) {
+        // Create a state transfer manager.
+        StateTransferManager transferManager =
+                new StateTransferManager(streamLog, runtime.getParameters().getBulkReadSize());
 
-            Set<String> lowRedundancyServers = getNodesWithReducedRedundancy(layout, layoutSegmentToMergeTo);
+        // Create lambda that, given a runtime, creates the log unit clients map.
+        LogUnitClientMap map = (CorfuRuntime currentRuntime) -> {
+            Layout currentLayout = currentRuntime.getLayoutView().getLayout();
```

Comment:
can you pass currentLayout as a method parameter? It makes algorithm less error prone

Reply:
I would have pass two parameters then. I also use a runtime layout of a Corfu runtime to get the log unit client.